csrc/index_compute.cpp

```diff
@@ -398,6 +460,59 @@ void IndexCompute::handle(Split* split) {
   }
 }
 
+bool IndexCompute::isModuloInvalidUnswitchedIndex(
+    IterDomain* out_concrete_id,
+    Val* out_ind,
+    Val* inner_extent) const {
+  // If not in the unswitched domain map, this domain has no dependent
+  // unswitched domain
+  auto unswitched_domain_map_it = unswitched_domain_map_.find(out_concrete_id);
+  if (unswitched_domain_map_it == unswitched_domain_map_.end()) {
+    return false;
+  }
+
+  for (const auto& unswitched_domain_list : unswitched_domain_map_it->second) {
+    TORCH_INTERNAL_ASSERT(!unswitched_domain_list.empty());
+
+    // If the stride is a multiple of the inner extent, the leaf
+    // unswitched index remains to be a valid maximum index as the
+    // module by the inner extent will be just zero. More
+    // specifically, the index for this unswitched domain would be (x
+    // - 1) * extent_of_inner_domain_0 * extent_of_inner_domain_1
+    // ...., so if the stride component, i.e., the multiplication of all
+    // the inner extents is divisible by the merge inner extent, its
+    // contribution propagated to the inner path will be zero
+    Val* stride = out_concrete_id->fusion()->oneVal();
+    for (auto it = unswitched_domain_list.begin();
+         it != unswitched_domain_list.end() - 1;
+         ++it) {
+      IterDomain* inner_id = *it;
+      stride = IrBuilder::mulExpr(stride, getExtent(inner_id));
+    }
+    if (simplifyExpr(IrBuilder::modExpr(stride, inner_extent))->isZero()) {
+      continue;
+    }
+
+    // Also, if the total extent including the inner domains is a
+    // divisible factor of the inner extent, the contribution by the
+    // unswitched domain is guaranteed to be still the maximum when
+    // propagated to the inner path
+    Val* total_extent =
+        IrBuilder::mulExpr(stride, getExtent(unswitched_domain_list.back()));
+    if (simplifyExpr(IrBuilder::modExpr(inner_extent, total_extent))
+            ->isZero()) {
+      continue;
+    }
```

Comment:
Looks like there are two cases known to be safe, the first one correspond to `distributeDivisibleDivMod` in expr simplifier, which says `(i * 128 + j) % 4 = j % 4`, the second one correspond to `distributeGcdRemainderDivMod` in expr simplifier, which says `if j < 128, then (i * 128 + j) % 1024 = (i * 128) % 1024 + j`. For the first case, the assigned index for `unswitched_domain_list.back()` is `i`, and for the second case, the assigned index for `unswitched_domain_list.back()` is `j`. Am I understanding correct?

Reply:
I think that's correct.

Reply:
Could you add this note to the code? At least for me, if you say "the distributeDivisibleDivMod pattern", I would immediately get what it means, and any other way of explaining I would have to translate it in order to understand.

Reply:
Added